cli/internal/oauth/api/api.go

```diff
@@ -96,16 +96,33 @@ func tryDecodeOAuthError(resp *http.Response) error {
 // authenticated or we have reached the time limit for authenticating (based on
 // the response from GetDeviceCode).
 func (a API) WaitForDeviceToken(ctx context.Context, state State) (TokenResponse, error) {
-	ticker := time.NewTicker(state.IntervalDuration())
+	// Ticker for polling tenant for login – based on the interval
+	// specified by the tenant response.
+	ticker := time.NewTimer(state.IntervalDuration())
 	defer ticker.Stop()
-	timeout := time.After(state.ExpiryDuration())
+
+	// Timeout context, for quitting polling for user credentials if
+	// we reach the timeout specified by the tenant.
+	// Does not get canceled if the user cancels execution.
+	timeoutCtx, cancel := context.WithTimeout(
+		context.Background(), state.ExpiryDuration())
+	defer cancel()
+
+	// Context for calling getDeviceToken – gets cancelled whether
+	// the caller cancels login or we hit the timeout.
+	getDeviceTokenCtx, getDeviceTokenCancel := context.WithTimeout(
+		ctx, state.ExpiryDuration())
```

Comment:
Wondering why we want to use separate contexts for these 🤔 


i.e.; wouldn't it be OK to cancel whatever we're doing if the parent context is canceled or has a timeout and times out? Basically, a cancelled context would mean; "stop what you're doing; we're no longer interested", and not sure if it (really?) matters whether that was because the parent context had a timeout set (and timed out), or the parent context was cancelled (the error message probably would indicate which of those was the case)?


If we _do_ want a separate `timeoutCtx`, should that one be using `context.WithoutCancel(ctx)` (i.e. still be based on the parent context, but removing the timeout)?

Reply:
We don't need to use the caller's ctx as a `timeoutCtx` parent as we don't really pass it anywhere. And the `getDeviceTokenCtx` also uses the same expiration.

If we really want to avoid context here, we can just simplify it to:

```go
func (a API) WaitForDeviceToken(ctx context.Context, state State) (TokenResponse, error) {
    expired := time.After(state.ExpiryDuration())
```

and then..
```go
	select {
		case <-ctx.Done():
			// user canceled login
			return TokenResponse{}, ctx.Err()
        case <-expired:
            return TokenResponse{}, ErrTimeout
```

Reply:
By selecting on multiple things that timeout at the same time, we're introducing a race condition and when the timeout is hit, `select` will randomly select a branch and we won't know if it happened because the caller cancelled the context or it timed out.

That's why I initially had it more like @vvoland's suggestion.

Reply:
However, @thaJeztah wanted put a timer on the context passed to `a.getDeviceToken`, which is why I introduced `getDeviceTokenCtx`.

Reply:
In the previous version we didn't `select` on `getDeviceTokenCtx` though, so IIUC we could make it not racy:

When both `timeoutCtx` and `getDeviceTokenCtx` would be `Done()` due to a deadline exceeded, it would either:

1. Immediately go to `case <-timeoutCtx.Done():` branch in case the timeout happened before `getDeviceToken` was called
2. `res, err := a.getDeviceToken(getDeviceTokenCtx, state)` would error out - in this case we could check if `timeoutCtx` is also `Done()` and either explicitly return `ErrTimeout` or just continue to the next loop iteration so it goes back into the 1.

Reply:
Yeah, my thinking was that `a.getDeviceToken` could hang, and in that case the login could hang forever?

```patch
git diff
diff --git a/cli/internal/oauth/api/api.go b/cli/internal/oauth/api/api.go
index 9ea48fecf..ff46126be 100644
--- a/cli/internal/oauth/api/api.go
+++ b/cli/internal/oauth/api/api.go
@@ -103,7 +103,7 @@ func (a API) WaitForDeviceToken(ctx context.Context, state State) (TokenResponse
        // The tenant tells us for as long as we can poll it for credentials
        // while the user logs in in the browser. Timeout if we don't get
        // credentials within this period.
-       timeout := time.NewTimer(state.ExpiryDuration())
+       timeout := time.NewTimer(10 * time.Second)
        defer timeout.Stop()

        for {
@@ -163,6 +163,7 @@ func (a API) getDeviceToken(ctx context.Context, state State) (TokenResponse, er
                "device_code": {state.DeviceCode},
        }
        oauthTokenURL := a.TenantURL + "/oauth/token"
+       time.Sleep(30 * time.Second)

        resp, err := postForm(ctx, oauthTokenURL, strings.NewReader(data.Encode()))
        if err != nil {
```



```bash
time ./build/docker login
...
Waiting for authentication in the browser…
failed waiting for authentication: timed out waiting for device token

________________________________________________________
Executed in   35.55 secs      fish           external
   usr time   26.23 millis    0.22 millis   26.02 millis
   sys time   52.20 millis    1.45 millis   50.74 millis
```

That said, `state.ExpiryDuration()` looks to be really long already (15 minutes) - slightly curious if that's really how long we want to wait for the user to log in :thinking_face:;

```
IntervalDuration: 5s
ExpiryDuration: 15m0s
```

But regardless, we could consider setting a timeout on the context so that the request would never take longer than the timeout;

```patch
git diff
diff --git a/cli/internal/oauth/api/api.go b/cli/internal/oauth/api/api.go
index 9ea48fecf..d8e2ea8bc 100644
--- a/cli/internal/oauth/api/api.go
+++ b/cli/internal/oauth/api/api.go
@@ -106,6 +106,9 @@ func (a API) WaitForDeviceToken(ctx context.Context, state State) (TokenResponse
        timeout := time.NewTimer(state.ExpiryDuration())
        defer timeout.Stop()

+       ctx, cancel := context.WithTimeout(ctx, state.ExpiryDuration())
+       defer cancel()
+
        for {
                resetTimer(ticker, state.IntervalDuration())
                select {
```

Not sure if we want the error message to be as-is (TBD?);

```
Waiting for authentication in the browser…
failed waiting for authentication: failed to get tokens: Post "https://login.docker.com/oauth/token": context deadline exceeded
```

Reply:
> slightly curious if that's really how long we want to wait for the user to log in :thinking_face:

Why not? The user can always cancel it. I don't think we should be trying to be smarter than the user here.

Reply:
> But regardless, we could consider setting a timeout on the context so that the request would never take longer than the timeout;

That's what I had previously 😅

Reply:
The current approach seems fairly simple, and is my preferred approach. There's no reason to replace the current `timeout` timer with a context afaict, other than to just avoid timers – but I think in this case replacing it with a context makes this less understandeable, not more.

We can also, additionally, set a reasonable deadline (e.g. 1 minute) inside `API.getDeviceToken`, to prevent it being stuck forever. Does that sound okay for all?

Reply:
Sounds good to me!

Reply:
Done ✅ @thaJeztah this LGTY?